piker/brokers/binance.py

```diff
@@ -198,12 +228,29 @@ async def bars(
             }
         )
 
-        new_bars = [
-            (i,) + tuple(
-                ftype(bar[j])
-                for j, (name, ftype) in enumerate(_ohlc_dtype[1:])
-            ) for i, bar in enumerate(bars)
-        ] 
+        # TODO: pack this bars scheme into a ``pydantic`` validator type:
+        # https://binance-docs.github.io/apidocs/spot/en/#kline-candlestick-data
+
+        # TODO: we should port this to ``pydantic`` to avoid doing
+        # manual validation ourselves..
+        new_bars = []
+        for i, bar in enumerate(bars):
+
+            bar = OHLC(*bar)
+
+            row = []
+            for j, (name, ftype) in enumerate(_ohlc_dtype[1:]):
+
+                # TODO: maybe we should go nanoseconds on all
+                # history time stamps?
+                if name == 'time':
+                    # convert to epoch seconds: float
+                    row.append(bar.time / 1000.0)
+
+                else:
+                    row.append(getattr(bar, name))
```

Comment:
This felt really really clunky.

What's a better way to one shot validate these?
I don't know if you can unpack `*bar` in a `pydantic` mode..

Reply:
Yeah I mucked with this for a while and got nowhere too fancy 😂 

Since our dtype is technically matching the dataclass we could actually probably just use `asdict` if we wanted to do it this way:

```python
[nav] In [116]: @dataclass
           ...: class C:
           ...:     doggy: int
           ...:     num: int

[nav] In [117]: l = [C(doggy=9, num=10), C(doggy=11, num=12)]

[nav] In [118]: np.array(list(list(asdict(c).values()) for c in l), dtype={'names': ('doggy', 'num',), 'formats': (int, int)})
Out[118]:
array([[( 9,  9), (10, 10)],
       [(11, 11), (12, 12)]], dtype=[('doggy', '<i8'), ('num', '<i8')])
```

Reply:
Using a `pydantic.BaseModel.dict()` would require array unpacking each `bar` into dict form since I still don't think `pydantic.BaseModel` supports the `OHLC(*bar)` style creation.

Not sure it'd be worth it anyway for readability.